### ISSUE TITLE
Weird block of RNN in TF2.2

### ISSUE THREAD

**Issue by BlueFisher** [tensorflow/tensorflow#39697]
**System information**
This bug exists in TF v2.2.0-rc4-8-g2b96f3662b 2.2.0 and on windows, linux, both CPU and GPU version. However, it does not exist in TF 2.1.

**Describe the current behavior**
Just check the code
```python
rnn = tf.keras.layers.GRU(3)
rnn(tf.keras.Input([None, 2]), tf.keras.Input([3, ]))


@tf.function(input_signature=[tf.TensorSpec(shape=(None, None, 2)),
                              tf.TensorSpec(shape=(None, 3))])
def test(x, i):
    with tf.GradientTape(persistent=True) as tape:
        rnn(x, initial_state=i)


test(np.random.randn(3, 2, 2).astype(np.float32),
     np.random.randn(3, 3).astype(np.float32))
```

The code will block in `test`, and the memory will grow unlimitedly.

The code will also block if `GRU` is replaced by `LSTM`. But it will run well if it is `SimpleRNN`.

This bug is a little complicated bacause when only all `rnn(tf.keras.Input...`, `@tf.function(input_signature=...)` and `with tf.GradientTape(persistent=True)...` exist, the block shows up.

**Comment by ravikyram** [comment]
I have tried in colab with TF version 2.1 , 2.2 and was able to reproduce the issue.Please, find the gist [here](https://colab.sandbox.google.com/gist/ravikyram/48ae035e8dd66abf4ea0350827b5cc29/untitled910.ipynb).Thanks!

**Comment by BlueFisher** [comment]
@ravikyram 👍

**Comment by qlzh727** [comment]
Thanks for reporting the issue. I think there is some regression/corner case in the tf.function. I will check with core team first and rely.

**Comment by K9-2-PE** [comment]
Any updates or potential work-arounds for this issue?  For my research I need intermediate jacobian values for use in loss calculation, and assume i need a persistent tape for jacobian wrt input columns and loss wrt model trainable variables.  Can I accomplish the same with nesting of some kind?

**Comment by BlueFisher** [comment]
@saikumarchalla It runs perfectly. Problem finally solved. Thanks!
